src/treesitter.rs

```diff
@@ -21,3 +25,103 @@ impl From<&'_ Node<'_>> for Match {
         Self::new(range.start_byte, range.end_byte)
     }
 }
+
+pub trait Parseable {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree>;
+}
+
+impl<'a> Parseable for &'a [u8] {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree> {
+        parser.parse(self, old_tree)
+    }
+}
+
+impl<'a> Parseable for &'a Rope {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree> {
+        parser.parse_with(
+            &mut |byte_offset, _| {
+                let (chunk, chunk_start_byte_index, _, _) = self.chunk_at_byte(byte_offset);
+                &chunk[byte_offset - chunk_start_byte_index..]
+            },
+            old_tree,
+        )
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum RopeOrSlice<'a> {
+    Slice(&'a [u8]),
+    Rope(&'a Rope),
+}
+
+impl<'a> TextProvider<'a> for RopeOrSlice<'a> {
+    type I = RopeOrSliceTextProviderIterator<'a>;
+
+    fn text(&mut self, node: Node) -> Self::I {
+        match self {
+            Self::Slice(slice) => {
+                RopeOrSliceTextProviderIterator::Slice(iter::once(&slice[node.byte_range()]))
```

Comment:
I actually wasn't clever enough to think to express this as an `iter::once()` myself, I initially wrote it with a manual "have we returned the one slice yet" state field but then I saw this is how they had it written somewhere (I guess in the baked-in `impl TextProvider for &[u8]`?)